sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SideInputLoadTest.java

```diff
@@ -0,0 +1,243 @@
+  private static final Instant TIME = new Instant();
+
+  public SideInputLoadTest(String[] args) throws IOException {
+    super(args, Options.class, METRICS_NAMESPACE);
+  }
+
+  @Override
+  void loadTest() throws IOException {
+    Optional<SyntheticStep> syntheticStep = createStep(options.getStepOptions());
+    PCollection<KV<byte[], byte[]>> input =
+        pipeline
+            .apply(readFromSource(sourceOptions))
+            .apply(ParDo.of(new AddTimestamps()))
+            .apply("Collect start time metrics", ParDo.of(runtimeMonitor))
+            .apply(ParDo.of(new ByteMonitor(METRICS_NAMESPACE, "totalBytes.count")));
+
+    performTestWithSideInput(
+        input, SideInputMaterializationType.valueOf(options.getSideInputType()), syntheticStep);
+  }
+
+  private void performTestWithSideInput(
+      PCollection<KV<byte[], byte[]>> input,
+      SideInputMaterializationType sideInputType,
+      Optional<SyntheticStep> syntheticStep) {
+    switch (sideInputType) {
+      case ITERABLE:
+        performTestWithIterable(input, syntheticStep);
+        break;
+      case HASHMAP:
+        performTestWithHashMap(input, syntheticStep);
+        break;
+      case LIST:
+        performTestWithList(input, syntheticStep);
+        break;
+    }
+  }
+
+  private void performTestWithList(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<List<KV<byte[], byte[]>>> sideInput = input.apply(View.asList());
+    input
+        .apply(ParDo.of(new SideInputTestWithList(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithHashMap(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Map<byte[], byte[]>> sideInput = input.apply(View.asMap());
+    input
+        .apply(ParDo.of(new SideInputTestWithHashMap(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
+  }
+
+  private void performTestWithIterable(
+      PCollection<KV<byte[], byte[]>> input, Optional<SyntheticStep> syntheticStep) {
+    applyStepIfPresent(input, "Synthetic step", syntheticStep);
+    PCollectionView<Iterable<KV<byte[], byte[]>>> sideInput;
+    if (options.getWindowCount() != 0) {
+      long windowDurationMilis = sourceOptions.numRecords / options.getWindowCount();
+      sideInput =
+          input
+              .apply(Window.into(FixedWindows.of(Duration.millis(windowDurationMilis))))
+              .apply(View.asIterable());
+    } else {
+      sideInput = input.apply(View.asIterable());
+    }
+
+    input
+        .apply(ParDo.of(new SideInputTestWithIterable(sideInput)).withSideInputs(sideInput))
+        .apply("Collect end time metrics", ParDo.of(runtimeMonitor));
```

Comment:
I just noticed `runtimeMonitor` is a `protected` field. It seems there are actually a few, and you have built enough on top that it may be getting hard to change. But even though I am late, I offer a suggestion for refactor: make `LoadTest` parameterized instead of a base class.

Each subclass of `LoadTest` is really just the main `loadTest` function and really that function is about the same as writing a `PTransform`. If you eliminate subclassing you can have `LoadTest` own the pipeline setup with so it will never be possible to forget or mess up `readSourceFromOptions` and `ParDo.of(runtimeMonitor)`. It will be less repeat boilerplate. Also you won't have to reference variables defined in another file as protected fields.

Reply:
I think we can talk about refactoring of the load tests in another context, let's keep this PR focused on the load test. WDYT?

Reply:
I see that one option is to create Jira with list of improvements and work on it in next PR. This will make it easier to review both cases.

Reply:
@kennknowles how do you envision this refactor? I don't see how we can parametrize the LoadTest further (it's already parametrized with options class). What we can do is move some of the boilerplate up to the LoadTest class, so that some of source reading is done automatically.

I will be grateful for your input in this matter.